tests/python/gstaichi/lang/fast_caching/test_src_ll_cache.py

```diff
@@ -139,3 +144,69 @@ def k1() -> None:
         assert cache_used == src_ll_cache
     else:
         assert cache_used  # default
+
+
+class TemplateParamsKernelArgs(pydantic.BaseModel):
+    arch: str
+    offline_cache_file_path: str
+    a: int
+    src_ll_cache: bool
+
+
+def src_ll_cache_template_params_child(args: list[str]) -> None:
+    print("src_ll_cache_template_params_child")
+    print("args", args)
+    args_obj = TemplateParamsKernelArgs.model_validate_json(args[0])
+    ti.init(
+        arch=getattr(ti, args_obj.arch),
+        offline_cache=True,
+        offline_cache_file_path=args_obj.offline_cache_file_path,
+        src_ll_cache=args_obj.src_ll_cache,
+    )
+    print("after init")
+
+    @ti.pure
+    @ti.kernel
+    def k1(a: ti.template(), output: ti.types.NDArray[ti.i32, 1]) -> None:
+        output[0] = a
+
+    print("after kernel declare")
+    output = ti.ndarray(ti.i32, (10,))
+    print("after create output. running kernel...")
+    k1(args_obj.a, output)
+    print("after running kernel")
+    assert output[0] == args_obj.a
+    print(TEST_RAN)
+
+
+@pytest.mark.parametrize("src_ll_cache", [False, True])
+@test_utils.test()
+def test_src_ll_cache_template_params(tmp_path: pathlib.Path, src_ll_cache: bool) -> None:
+    """
+    template primitive kernel params should be in the cache key
+    """
+    arch = ti.lang.impl.current_cfg().arch.name
+
+    def create_args(a: int) -> str:
+        obj = TemplateParamsKernelArgs(
+            arch=arch,
+            offline_cache_file_path=str(tmp_path),
+            src_ll_cache=src_ll_cache,
+            a=a,
+        )
+        json = TemplateParamsKernelArgs.model_dump_json(obj)
+        return json
+
+    env = os.environ
+    env["PYTHONPATH"] = "."
+    for a in [3, 4]:
+        assert TEST_RAN in subprocess.check_output(
+            [sys.executable, __file__, src_ll_cache_template_params_child.__name__, create_args(a)]
+        ).decode("utf-8")
+
+
+# The following lines are critical for the tests to work. If they are missing, the test will
+# incorrectly pass, without doing anything.
+if __name__ == "__main__":
+    print("__main__")
+    globals()[sys.argv[1]](sys.argv[2:])
```

Comment:
I would recommend to rather use a specific exit code in case of success. I would suggest 42:
https://unix.stackexchange.com/a/604262/406277

Reply:
I'd prefer to use text output on the whole, because it's more human friendly I feel.

I think the main advantage of using a return code would be performance? But for these tests, the performance gain here would be minimal.

Or ... what do you see as advantages to using a magic return code?

Reply:
Added return code check